File: Objective-C Demo App/iOS-SDK-Demo/ALDemoRootViewController.m
//
//  ALDemoRootViewController.m
//  iOS-SDK-Demo
//
//  Created by Thomas So on 9/23/15.
//  Copyright © 2015 AppLovin. All rights reserved.
//

@import AppLovinSDK;
#import "ALDemoRootViewController.h"
#import <MessageUI/MFMailComposeViewController.h>
#import <SafariServices/SafariServices.h>

@interface ALDemoRootViewController()<MFMailComposeViewControllerDelegate>
@property (nonatomic, strong) IBOutlet UIBarButtonItem *muteToggle;
@end

@implementation ALDemoRootViewController
static NSString *const kSupportEmail = @"support@applovin.com";
static NSString *const kSupportLink = @"https://support.applovin.com/support/home";

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self.navigationController setToolbarHidden: YES];
    
    UIImageView *logo = [[UIImageView alloc] initWithFrame: CGRectMake(0.0f, 0.0f, 40.0f, 40.0f)];
    logo.contentMode = UIViewContentModeScaleAspectFit;
    logo.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
    logo.image = [UIImage imageNamed: @"nav_logo"];
    self.navigationItem.titleView = logo;
    
    [self addFooterLabel];
    
    self.muteToggle.image = [self muteIconForCurrentSdkMuteSetting];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear: animated];

    // Re-set status bar style after returning from SFSafariViewController
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
    [[UIApplication sharedApplication] setStatusBarStyle: UIStatusBarStyleLightContent];
#pragma GCC diagnostic pop
}

#pragma mark - Table View Delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath: indexPath animated: YES];
    
    if ( indexPath.section == 1 )
    {
        if ( indexPath.row == 0 )
        {
            [self openSupportSite];
        }
        else if ( indexPath.row == 1 )
        {
            [self attemptSendEmail];
        }
    }
}

#pragma mark - Sound Toggling

- (IBAction)toggleMute:(UIBarButtonItem *)sender
{
    /**
     * Toggling the sdk mute setting will affect whether your video ads begin in a muted state or not.
     */
    ALSdk *sdk = [ALSdk shared];
    sdk.settings.muted = !sdk.settings.muted;
    sender.image = [self muteIconForCurrentSdkMuteSetting];
}

- (UIImage *)muteIconForCurrentSdkMuteSetting
{
    return [ALSdk shared].settings.muted ? [UIImage imageNamed: @"mute"] : [UIImage imageNamed: @"unmute"];
}

#pragma mark - Table View Actions

- (void)openSupportSite
{
    NSOperatingSystemVersion version = [[NSProcessInfo processInfo]operatingSystemVersion];
    if ( version.majorVersion > 8 )
    {
        SFSafariViewController *safariController = [[SFSafariViewController alloc] initWithURL: [NSURL URLWithString: kSupportLink]
                                                                       entersReaderIfAvailable: YES];
        [self presentViewController: safariController animated: YES completion:^{
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
            [[UIApplication sharedApplication] setStatusBarStyle: UIStatusBarStyleDefault];
#pragma GCC diagnostic pop
        }];
    }
    else
    {
        [[UIApplication sharedApplication] openURL: [NSURL URLWithString: kSupportLink]];
    }
}

- (void)attemptSendEmail
{
    if ( [MFMailComposeViewController canSendMail] )
    {
        MFMailComposeViewController *mailController = [[MFMailComposeViewController alloc] init];
        mailController.mailComposeDelegate = self;
        [mailController setSubject: @"iOS SDK support"];
        [mailController setToRecipients: @[kSupportEmail]];
        [mailController setMessageBody: [NSString stringWithFormat: @"\n\n---\nSDK Version: %@", [ALSdk version]] isHTML: NO];
        mailController.navigationBar.tintColor = [UIColor whiteColor];
        [self presentViewController: mailController animated: YES completion:^{
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
            [[UIApplication sharedApplication] setStatusBarStyle: UIStatusBarStyleLightContent];
#pragma GCC diagnostic pop
        }];
    }
    else
    {
        NSString *message = [NSString stringWithFormat: @"Your device is not configured for sending emails.\n\nPlease send emails to %@", kSupportEmail];
        [[[UIAlertView alloc] initWithTitle: @"Email Unavailable"
                                    message: message
                                   delegate: nil
                          cancelButtonTitle: @"OK"
                          otherButtonTitles: nil] show];
    }
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    switch ( result )
    {
        case MFMailComposeResultSent:
            [[[UIAlertView alloc] initWithTitle: @"Email Sent"
                                        message: @"Thank you for your email, we will process it as soon as possible."
                                       delegate: nil
                              cancelButtonTitle: @"OK"
                              otherButtonTitles: nil] show];
        case MFMailComposeResultCancelled:
        case MFMailComposeResultSaved:
        case MFMailComposeResultFailed:
        default:
            break;
    }
    
    [self dismissViewControllerAnimated: YES completion: nil];
}

- (void)addFooterLabel
{
    UILabel *footer = [[UILabel alloc] init];
    footer.font = [UIFont systemFontOfSize: 14.0f];
    footer.numberOfLines = 0;
    
    NSString *appVersion = [[NSBundle mainBundle] infoDictionary][@"CFBundleShortVersionString"];
    NSString *sdkVersion = [ALSdk version];
    NSString *systemVersion = [[UIDevice currentDevice] systemVersion];
    NSString *text = [NSString stringWithFormat: @"App Version: %@\nSDK Version: %@\niOS Version: %@\n\nLanguage: Objective-C", appVersion, sdkVersion, systemVersion];
    
    NSMutableParagraphStyle *style  = [[NSMutableParagraphStyle alloc] init];
    style.alignment =  NSTextAlignmentCenter;
    style.minimumLineHeight = 20.0f;
    footer.attributedText = [[NSAttributedString alloc] initWithString: text attributes: @{NSParagraphStyleAttributeName : style}];
    
    CGRect frame = footer.frame;
    frame.size.height = [footer sizeThatFits: CGSizeMake(CGRectGetWidth(footer.frame), CGFLOAT_MAX)].height + 60.0f;
    footer.frame = frame;
    self.tableView.tableFooterView = footer;
}

@end

File: Swift Demo App/iOS-SDK-Demo/ALDemoRootViewController.swift
//
//  ALDemoRootViewController.swift
//  iOS-SDK-Demo
//
//  Created by Thomas So on 9/25/15.
//  Copyright © 2015 AppLovin. All rights reserved.
//

import UIKit
import MessageUI
import SafariServices

class ALDemoRootViewController: UITableViewController, MFMailComposeViewControllerDelegate
{
    let kSupportEmail = "support@applovin.com"
    let kSupportLink = "https://support.applovin.com/support/home"

    let kRowIndexToHideForPhones = 5;
    
    @IBOutlet var muteToggle: UIBarButtonItem!
    
    // MARK: View Lifecycle
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        self.navigationController?.setToolbarHidden(true, animated: false)
        
        let logo = UIImageView(image: UIImage(named: "nav_logo"))
        logo.frame = CGRect(x: 0, y: 0, width: 40, height: 40)
        logo.contentMode = .scaleAspectFit
        self.navigationItem.titleView = logo
        
        self.addFooterLabel()
        
        self.muteToggle.image = muteIconForCurrentSdkMuteSetting()
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
        
        // Re-set status bar style after returning from SFSafariViewController
        UIApplication.shared.statusBarStyle = .lightContent
    }
    
    override func viewDidAppear(_ animated: Bool)
    {
        super.viewDidAppear(animated)
        
        let sdkKey = Bundle.main.infoDictionary!["AppLovinSdkKey"] as! String
        if sdkKey == "YOUR_SDK_KEY"
        {
            let alertVC = UIAlertController(title: "ERROR", message: "Please update the `AppLovinSdkKey` row in your Info.plist file with your SDK key.", preferredStyle: .alert)
            let okAction = UIAlertAction(title: "OK", style: .cancel)
            alertVC.addAction(okAction)
            self.present(alertVC, animated: true, completion: nil)
        }
    }
    
    // MARK: Table View Delegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath)
    {
        tableView.deselectRow(at: indexPath, animated: true)
        
        if indexPath.section == 1
        {
            if indexPath.row == 0
            {
                openSupportSite()
            }
            else if indexPath.row == 1
            {
                attemptSendEmail()
            }
        }
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath)
    {
        if UIDevice.current.userInterfaceIdiom == .phone && indexPath.section == 0 && indexPath.row  == kRowIndexToHideForPhones
        {
            cell.isHidden = true;
        }
    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat
    {
        if UIDevice.current.userInterfaceIdiom == .phone && indexPath.section == 0 && indexPath.row  == kRowIndexToHideForPhones
        {
            return 0;
        }

        return super.tableView(tableView, heightForRowAt: indexPath)
    }

    // MARK: Sound Toggling

    @IBAction func toggleMute(_ sender: UIBarButtonItem!)
    {
        /**
         * Toggling the sdk mute setting will affect whether your video ads begin in a muted state or not.
         */
        let sdk = ALSdk.shared()
        sdk?.settings.muted = !(sdk?.settings.muted)!
        sender.image = muteIconForCurrentSdkMuteSetting()
    }
    
    func muteIconForCurrentSdkMuteSetting() -> UIImage!
    {
        return ALSdk.shared()!.settings.muted ? UIImage(named: "mute") : UIImage(named: "unmute")
    }
    
    // MARK: Table View Actions
    
    func openSupportSite()
    {
        guard let supportURL = URL(string: kSupportLink) else { return }
        
        if #available(iOS 9.0, *)
        {
            let safariController = SFSafariViewController(url: supportURL, entersReaderIfAvailable: true)
            self.present(safariController, animated: true, completion: {
                UIApplication.shared.statusBarStyle = .default
            })
        }
        else
        {
            UIApplication.shared.openURL(supportURL)
        }
    }
    
    func attemptSendEmail()
    {
        if MFMailComposeViewController.canSendMail()
        {
            let mailController = MFMailComposeViewController()
            mailController.mailComposeDelegate = self
            mailController.setSubject("iOS SDK Support")
            mailController.setToRecipients([kSupportEmail])
            mailController.setMessageBody("\n\n---\nSDK Version: \(ALSdk.version())", isHTML: false)
            mailController.navigationBar.tintColor = UIColor.white
            
            self.present(mailController, animated: true, completion: {
                UIApplication.shared.statusBarStyle = .lightContent
            })
        }
        else
        {
            let message = "Your device is not configured for sending emails.\n\nPlease send emails to \(kSupportEmail)"
            let alertVC = UIAlertController(title: "Email Unavailable", message: message, preferredStyle: .alert)
            let okAction = UIAlertAction(title: "OK", style: .cancel)
            alertVC.addAction(okAction)
            self.present(alertVC, animated: true, completion: nil)
        }
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?)
    {
        switch ( result.rawValue )
        {
        case ( MFMailComposeResult.sent.rawValue ):
            let alertVC = UIAlertController(title: "Email Sent", message: "Thank you for your email, we will process it as soon as possible.", preferredStyle: .alert)
            let okAction = UIAlertAction(title: "OK", style: .cancel)
            alertVC.addAction(okAction)
            self.present(alertVC, animated: true, completion: nil)
        default:
            break
        }
        
        self.dismiss(animated: true, completion: nil)
    }
    
    func addFooterLabel()
    {
        let footer = UILabel()
        footer.font = UIFont.systemFont(ofSize: 14)
        footer.numberOfLines = 0
        
        let appVersion = Bundle.main.infoDictionary!["CFBundleShortVersionString"] as! String
        let sdkVersion = ALSdk.version()
        let systemVersion = UIDevice.current.systemVersion
        let text = "App Version: \(appVersion)\nSDK Version: \(sdkVersion)\niOS Version: \(systemVersion)\n\nLanguage: Swift"
        
        let style = NSMutableParagraphStyle()
        style.alignment = .center
        style.minimumLineHeight = 20
        footer.attributedText = NSAttributedString(string: text, attributes: [NSAttributedString.Key.paragraphStyle : style])
        
        var frame = footer.frame
        frame.size.height = footer.sizeThatFits(CGSize(width: footer.frame.width, height: CGFloat.greatestFiniteMagnitude)).height + 60
        footer.frame = frame
        self.tableView.tableFooterView = footer
    }
}
